jetstream_consumer.go

```diff
@@ -0,0 +1,333 @@
+// Copyright 2020 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nats
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+)
+
+func (nc *Conn) createOrUpdateConsumer(opts *jsOpts) (*ConsumerInfo, error) {
+	if opts.streamName == "" {
+		return nil, ErrStreamNameRequired
+	}
+	if opts.consumer == nil {
+		return nil, ErrConsumerConfigRequired
+	}
+
+	crj, err := json.Marshal(&JSApiConsumerCreateRequest{
+		Stream: opts.streamName,
+		Config: *opts.consumer,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := opts.context(nc.Opts.JetStreamTimeout)
+	defer cancel()
+
+	var subj string
+	switch len(opts.consumer.Durable) {
+	case 0:
+		subj = fmt.Sprintf(JSApiConsumerCreateT, opts.streamName)
+	default:
+		subj = fmt.Sprintf(JSApiDurableCreateT, opts.streamName, opts.consumer.Durable)
+	}
+
+	resp, err := nc.RequestWithContext(ctx, subj, crj)
+	if err != nil {
+		return nil, err
+	}
+
+	cresp := &JSApiConsumerCreateResponse{}
+	err = json.Unmarshal(resp.Data, cresp)
+	if err != nil {
+		return nil, err
+	}
+
+	if cresp.IsError() {
+		return nil, cresp.ToError()
+	}
+
+	return cresp.ConsumerInfo, nil
+}
+
+const (
+	JSApiConsumerCreateT = "$JS.API.CONSUMER.CREATE.%s"
+	JSApiDurableCreateT  = "$JS.API.CONSUMER.DURABLE.CREATE.%s.%s"
+)
+
+type ApiError struct {
+	Code        int    `json:"code"`
+	Description string `json:"description,omitempty"`
+}
+
+// Error implements error
+func (e ApiError) Error() string {
+	switch {
+	case e.Description == "" && e.Code == 0:
+		return "unknown JetStream Error"
+	case e.Description == "" && e.Code > 0:
+		return fmt.Sprintf("unknown JetStream %d Error", e.Code)
+	default:
+		return e.Description
+	}
+}
+
+// NotFoundError is true when the error is one about a resource not found
+func (e ApiError) NotFoundError() bool { return e.Code == 404 }
```

Comment:
I still think in clients we absorb this and don't surface, just do new Error for each one where it may surface. This feels like jsm not base client lib.

Reply:
not used here so just removed them, we only use the thing that makes a err from a response.  I copy/pasted all the code rather than cherry pick since while we are still changing how things work it will be a pain to keep yet-another-package up to date, but fine, will do